server/apple_notification_server.go

```diff
@@ -269,3 +262,56 @@ func (me *AppleNotificationServer) SendNotification(msg *PushNotification) PushR
 	}
 	return NewOkPushResponse()
 }
+
+func (me *AppleNotificationServer) SendNotificationWithRetry(notification *apns.Notification) (*apns.Response, error) {
+	var res *apns.Response
+	var err error
+	waitTime := time.Second
+
+	// Keep a general context to make sure the whole retry
+	// doesn't take longer than the timeout.
+	generalContext, cancelGeneralContext := context.WithTimeout(context.Background(), me.sendTimeout)
+	defer cancelGeneralContext()
+
+	// Set the retry context timeout to a value that allow us to retry the notification
+	// the MAX RETRIES with exponential backoff. With default values, this timeout
+	// will be 5 seconds.
+	retryContextTimeout := me.sendTimeout / (MAX_RETRIES * 2)
```

Comment:
The retry value would be 60 secs? So the timeout for the context here is 10 secs ?

Reply:
`me.sendTimeout` is by default 30s. `MAX_RETRIES` is 3.

30 / (3* 2) = 5 seconds

So... worst case scenario with the defaults...
- 5 seconds trying the first time and fail because the context exceeded the deadline
- Wait 1 second
- 5 seconds trying the second time and fail because the context exceeded the deadline
- Wait 2 seconds
- 5 seconds trying the third time and fail because the context exceeded the deadline
- Return error

A total of 18s, more than enough to fit in the 30s timeout.

I can do something more hardcoded, or tweak the numbers if we prefer. I don't have a strong opinion on these.

Reply:
Are we confident that 5 seconds is enough to set the timeout for the apple client call? It seems we are going from 30 secs per call to only 5 secs per call. In the dashboard we saw there were times it would take up to 10 secs.

Just trying to understand why we are trying to fix all the retry logic within the initial configuration of time out per send?

Reply:
Agree. The sendTimeout is the timeout for a single send. It should not be made to limit the time including retries.

Reply:
My worry is that 30 seconds 3 times is +90 seconds the server waiting for the push proxy response. The server problably has also a context deadline, so not sure if we want to wait that long.

But if you have a strong feeling we should just use the configured timeout for each retry, I am happy to do it.

Reply:
It's about semantics. The sendTimeout should be the timeout for a single request. If you want to have a separate timeout including all the retries, then that should be a separate config setting.

The server timeout is hardcoded to 30s currently.

Reply:
Added a new config, and use both to properly handle this. I set it to 8 seconds, since it is the highest integer we can use that will fit comfortably in the 30 seconds (8*3 + 1 + 2 = 27s).